### PR TITLE
fix(web): improve realtime subscriptions

### PR DESCRIPTION
**Pull request by TorstenDittmann**
- completely changes the way of how we manage callbacks
- instead of relying for each subscription on the native `WebSocket` event listener, we are handling them ourselves
- more quality of life improvements

output: https://gist.github.com/TorstenDittmann/5f8052cf800faa58db6897ec12b96a68

published: `npm i aw-test-realtime@4.1.1`

### CODE REVIEW

File: templates/web/src/sdk.ts.twig [hunk 1]
```diff
@@ -137,71 +143,103 @@ class {{ spec.title | caseUcfirst }} {
...
+
+                this.realtime.socket.addEventListener('close', event => {
+                    if (!this.realtime.reconnect || (this.realtime?.lastMessage?.type === 'error' && (<RealtimeResponseError>this.realtime?.lastMessage.data).code === 1008)) {
+                        this.realtime.reconnect = true;
+                        return;
+                    }
+                    console.error('Realtime got disconnected. Reconnect will be attempted in 1 second.', event.reason);
+                    setTimeout(() => {
+                        this.realtime.createSocket();
+                    }, 1000);
```
Comment (Meldiron) [inline templates/web/src/sdk.ts.twig:186]: On the weekend Eldad had an idea to slow it down gradually, can we do that? For example:

1st - 5th request: Every 1 second
5th - 15th request: Evety 5 seconds
15th - 100th request: Every 10 seconds
100th+ - Every 60 seconds

Comment (TorstenDittmann) [reply templates/web/src/sdk.ts.twig:186]: Oh yeah!

Forgot about that, going to add 🙏🏻

Comment (TorstenDittmann) [reply templates/web/src/sdk.ts.twig]: Added 👍🏻

**Review by Meldiron — COMMENTED** [review]
One comment

File: templates/web/src/sdk.ts.twig [hunk 2]
```diff
@@ -137,71 +145,122 @@ class {{ spec.title | caseUcfirst }} {
...
 
-            this.realtime.socket = new WebSocket(this.config.endpointRealtime + '/realtime?' + channels.toString());
-            this.realtime.socket?.addEventListener('message', this.realtime.authenticate);
+            const url = this.config.endpointRealtime + '/realtime?' + channels.toString();
 
-            for (const channel in this.realtime.channels) {
-                this.realtime.channels[channel].forEach(callback => {
-                    this.realtime.socket?.addEventListener('message', callback);
+            if (url !== this.realtime.url || !this.realtime.socket || this.realtime.socket?.readyState > WebSocket.OPEN) {
+                if (this.realtime.socket && this.realtime.socket?.readyState < WebSocket.CLOSING) {
```
Comment (christyjacob4) [inline templates/web/src/sdk.ts.twig:184]: Can we add some comments here to help understand these conditionals?  Kind of getting lost in here

File: templates/web/src/sdk.ts.twig [hunk 3]
```diff
@@ -137,71 +145,122 @@ class {{ spec.title | caseUcfirst }} {
...
+
+                this.realtime.url = url;
+                this.realtime.socket = new WebSocket(url);
+                this.realtime.socket.addEventListener('message', this.realtime.onMessage);
+                this.realtime.socket.addEventListener('open', event => {
+                    this.realtime.reconnectAttempts = 0;
                 });
-            }
+                this.realtime.socket.addEventListener('close', event => {
+                    if (!this.realtime.reconnect || (this.realtime?.lastMessage?.type === 'error' && (<RealtimeResponseError>this.realtime?.lastMessage.data).code === 1008)) {
```
Comment (christyjacob4) [inline templates/web/src/sdk.ts.twig:196]: Here as well. especially with the status code

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Left some minor comments. Rest looks good

**Review by christyjacob4 — COMMENTED** [review]
Looks great . thank you for adding those in :+1:

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**